Adding a package for testing lifecycle agent.

achuzhoy opened the pull request:
This commit adds support for testing IBU, a.k.a.
imagebasedupgrade.

achuzhoy commented:
cc @mcornea

cdvultur reviewed (commented):
Small comments. Not sure if really applicable

achuzhoy commented on pkg/lca/imagebasedupgrade.go:
```diff
@@ -0,0 +1,354 @@
+
+// WithSeedImage sets the seed image used by the imagebasedupgrade.
+func (builder *ImageBasedUpgrade) WithSeedImage(
+	seedImage string) *ImageBasedUpgrade {
```
It works setting to empty 

from `spec:                      
  additionalImages:                        
    name: ""
    namespace: ""  
  seedImageRef:
    image: quay.io/achuzhoy/ibu-seed-sno0:4.15.0-rc.1
    version: 4.15.0-rc.1
  stage: Idle
`
will be:
`
spec:                      
  additionalImages:                        
    name: ""
    namespace: ""  
  seedImageRef:
    image: quay.io/achuzhoy/ibu-seed-sno0:4.15.0-rc.1
    version: 4.15.0-rc.1
  stage: Idle
`

cdvultur replied:
So no checks are needed !

trewest replied:
@achuzhoy I think this is what we should be checking
```
spec:
  additionalImages:
    name: ""
    namespace: ""
  seedImageRef:
    image: ""
    version: 4.15.0-rc.1
  stage: Idle
```

achuzhoy replied:
```spec:
  additionalImages:
    name: ""
    namespace: ""
  seedImageRef:
    image: ""
    version: 4.15.0-rc.1
  stage: Idle
```

achuzhoy replied:
gets set - it will fail the upgrade though

achuzhoy commented on pkg/lca/imagebasedupgrade.go:
```diff
@@ -0,0 +1,354 @@
+
+// WithSeedImageVersion sets the seed image version used by the imagebasedupgrade.
+func (builder *ImageBasedUpgrade) WithSeedImageVersion(
+	seedImageVersion string) *ImageBasedUpgrade {
```
It works setting to empty 

from `spec:                      
  additionalImages:                        
    name: ""
    namespace: ""  
  seedImageRef:
    image: quay.io/achuzhoy/ibu-seed-sno0:4.15.0-rc.1
    version: 4.15.0-rc.1
  stage: Idle
`
will be:
`
spec:                      
  additionalImages:                        
    name: ""
    namespace: ""  
  seedImageRef:
    image: quay.io/achuzhoy/ibu-seed-sno0:4.15.0-rc.1
    version: 4.15.0-rc.1
  stage: Idle
`

trewest replied:
@achuzhoy I think this is what we should be checking
```
spec:
  additionalImages:
    name: ""
    namespace: ""
  seedImageRef:
    image: quay.io/achuzhoy/ibu-seed-sno0:4.15.0-rc.1
    version: ""
  stage: Idle
```

achuzhoy replied:
same comment above

cdvultur commented on pkg/lca/imagebasedupgrade.go:
```diff
@@ -0,0 +1,354 @@
+// WaitUntilStageTrue waits the specified timeout for the imagebasedupgrade to reach status true
+// for the provided stage .
+func (builder *ImageBasedUpgrade) WaitUntilStageTrue(
+	stage lcav1alpha1.ImageBasedUpgradeStage) (*ImageBasedUpgrade, error) {
```
Correct. But that is inside `WithStage` function.

In KMM there are strings being fed into like this: https://github.com/openshift-kni/eco-goinfra/blob/975bf2617254cf9ae66490e3a03b7307cbc6f272/pkg/kmm/kernelmapping.go#L25

achuzhoy replied:
I updated to strings.
Thanks.

trewest commented:
@achuzhoy I don't see any updates the clients pkg. Is that needed?

cdvultur reviewed (approved):
/lgtm

trewest commented:
> @achuzhoy I don't see any updates the clients pkg. Is that needed?

I also don't see any updates to go.mod and go.sum

achuzhoy commented:
> > > @achuzhoy I don't see any updates the clients pkg. Is that needed?
> > 
> > 
> > I also don't see any updates to go.mod and go.sum
> 
> @trewest I don't think go.mod and go.sum updates are needed given the repo already exists.
> 
> But on the other hand it requires update to the client. ( Comparing with KMM-HUB [#217 (files)](https://github.com/openshift-kni/eco-goinfra/pull/217/files) )
> 
> Good catch!
> 
> @achuzhoy there needs to be an update to clients also

Great cache - added.

achuzhoy commented:
> > @trewest I don't think go.mod and go.sum updates are needed given the repo already exists.
> 
> @cdvultur I'm not sure if I follow you. The lifecycle-agent module hasn't been added to our go.mod file from as far as I can tell. It's needed to ensure proper dependency management. @achuzhoy I'm assuming your go.mod and go.sum files have uncommited changes since you were able to run `go mod vendor`. Is that correct?



No, the only changed files in my testbed are:
```
[achuzhoy@achuzhoy eco-gotests]$ git status
On branch main
Changes to be committed:
  (use "git restore --staged <file>..." to unstage)
	modified:   tests/assisted/ztp/operator/operator_suite_test.go
	new file:   tests/assisted/ztp/operator/tests/lca_sasha.go
	modified:   vendor/github.com/openshift-kni/eco-goinfra/pkg/assisted/agentserviceconfig.go
	modified:   vendor/github.com/openshift-kni/eco-goinfra/pkg/clients/clients.go
	new file:   vendor/github.com/openshift-kni/eco-goinfra/pkg/lca/imagebasedupgrade.go
	new file:   vendor/github.com/openshift-kni/eco-goinfra/pkg/lca/seedgenerator.go_todo
	new file:   vendor/github.com/openshift-kni/lifecycle-agent/LICENSE
	new file:   vendor/github.com/openshift-kni/lifecycle-agent/api/v1alpha1/groupversion_info.go
	new file:   vendor/github.com/openshift-kni/lifecycle-agent/api/v1alpha1/types.go
	new file:   vendor/github.com/openshift-kni/lifecycle-agent/api/v1alpha1/zz_generated.deepcopy.go
```

achuzhoy commented:
cc @mpmaruthu

trewest reviewed (approved):
/lgtm

achuzhoy commented:
@cdvultur @trewest  you don't need to re-review the code after the recent push

I only changed one line  from:
condition.Type == "RollbackCompleted"
to:
condition.Type == "RollbackInProgress"

following an e2e tests

cdvultur reviewed (approved):
/lgtm

trewest reviewed (approved):
/lgtm
